File: ruby/lib/thermostat.rb
class Thermostat
  attr_reader :temperature

  DEFAULT_TEMPERATURE = 20

  def initialize
    @temperature = DEFAULT_TEMPERATURE
    @psm = true
  end

  def up
    @temperature += 1 unless maximum_reached?
  end

  def down
    @temperature -= 1 unless minimum_reached?
  end

  def reset
    @temperature = DEFAULT_TEMPERATURE
  end

  def self.instance
    @thermostat ||= self.new
  end

  def in_psm?
    @psm
  end

  def power_save_off
    @psm = false
  end

  def power_save_on
    @psm = true
    @temperature = 25 if maximum_reached?
  end

  def energy_usage
    if temperature < 18
      return :low
    elsif temperature < 25
      return :medium
    end
    return :high
  end

  private

  def maximum_reached?
    if in_psm?
      temperature >= 25
    else
      temperature == 32
    end
  end

  def minimum_reached?
    temperature == 10
  end

end

File: javascript/src/thermostat.js
'use strict';

class Thermostat{

    constructor(){
        this.DEFAULT_TEMP = 20;
        this.temp = this.DEFAULT_TEMP;
        this.MINIMUM_TEMP = 10;
        this.powerSaving = true;
        this.MAX_TEMP_PSM = 25;
        this.MAX_TEMP = 32;
        this.MEDIUM_USAGE = 18;
        this.HIGH_USAGE = 25;
    }

    getCurrentTemp(){
        return this.temp;
    }

    isMinTemp(){
        return this.temp === this.MINIMUM_TEMP;
    }

    isMaxTemp(){
        if (this.isPowerSavingOn() === false) {
            return this.temp === this.MAX_TEMP;
        }
        return this.temp === this.MAX_TEMP_PSM;
    }

    up(){
        if (this.isMaxTemp()) {
            return;   
        }
        this.temp++;
    }

    down(){
        if (this.isMinTemp()) {   
            return;
        }
        this.temp--;
    }

    isPowerSavingOn(){
        return this.powerSaving === true;
    }

    switchPowerSavingOff(){
        return this.powerSaving = false;
    }

    switchPowerSavingOn(){
        if (this.temp > this.MAX_TEMP_PSM){
            this.powerSaving = true;
            this.temp = this.MAX_TEMP_PSM;
        }
        return this.powerSaving = true;
    }

   
    reset(){
        this.temp = this.DEFAULT_TEMP;
    }

    getEnergyUsage(){
        if(this.temp < this.MEDIUM_USAGE) {
            return "LOW";
        } else if(this.temp <= this.HIGH_USAGE) {
            return "MEDIUM";
        }
            return "HIGH";
    }
}